### DISCUSSION TITLE
How to set a custom post type in faustjs to find the corresponding page?

### DISCUSSION THREAD

**Discussion by youdianfan** [wpengine/faustjs#1308] [Q&A]
I custom post address is http://localhost/events/brain-organoids-training-course

When I visit in faustjs and the page returns 404, how do I solve this problem?

![image](https://user-images.githubusercontent.com/27018910/222743671-ae0ed488-bab8-4022-a7ed-e9590f16f1dc.png)

**Comment by theodesp** [comment]
Hey. If your page does not resolve to  valid template type, you may need to add a custom hook. See the example in Portfolio Blueprint:

https://github.com/wpengine/atlas-blueprint-portfolio/blob/main/plugins/ProjectTemplatePlugin.js

```jsx
export class EventTemplatePlugin {
  constructor() {}

  apply(hooks) {
    hooks.addFilter('possibleTemplatesList', 'faust', (templates, data) => {
      if (data?.seedNode?.__typename === 'Event') {
        return Array.from(new Set(['event', ...templates]));
      }
      return templates;
    });
  }
}

```

**Comment by theodesp** [comment]
Behind the scenes Faust uses a seed query to try to find a valid template to use for each page:

https://faustjs.org/docs/faustwp/seed-query

In your example, if the path `/events/brain-organoids-training-course` returns `null` for the seedQuery then it won't match any template, so you need to manually add this to the `possibleTemplatesList`. It all depends on how the CPT was saved using `has_archive` I think

**Reply by slowdub** [reply]
@theodesp `possibleTemplatesList` hook doesn't even run for me if the request resolves to 404. So, there's really no opportunity at that time to fix the mismatched SeedQuery (mismatched as in returning `null`). Is this correct?

**Comment by brittonwalker** [comment]
I have the same problem. I'm using The Events Calendar which is registering a the post type and link path. I get a 404 and not sure how to get around this.

**Comment by iamkaranpatel** [comment]
I had the same issue when I was testing Faust js, I have created a post type **products** 

changes the filename to [[single-product]].js once done restart the localhost

hope this helps

**Comment by theodesp** [comment]
To see the `possibleTemplatesList`, you need to inspect the server logs in the console. When you visit a CPT page for example and having the above plugin running you should be able to see the info:
```
info - Possible templates for resolved url: "/events/brain-organoids-training-course": [ 'event', ... ]
```

This means that the seedQuery was able to resolve the page metadata. Then you need to create a template for the event CPT and export this in the index.js

```js
export default {
  'front-page': frontPage,
  page,
  single,
  event, // <<---- Event template
  archive,
};
```

If you instead do not see any Possible templates list, then it means that the seedQuery for the `/events/brain-organoids-training-course` uri resolved to `null`. This means that in your WordPress site you won't be able to resolve the same uri path.

**Comment by nazmateo** [comment]
Hey guys, I might have a same problem but it is reversed. I have posts that has specific categories. I have a single-post template. My problem is that for example, i have a post that has a category 'news' and a uri 'how-to'. Now I can view the content on /news/how-to path. But the problem is that even i replace the url in the browser to /customer/how-to, i can still view the page, basically it should give a 404 page for the /customer/how-to page.

Can someone help thank you. 


basically this is the graphql request


SingleNews.variables = ({ uri }) => {
  console.log(uri);
  return {
    uri,
    headerLocation: MENUS.PRIMARY_LOCATION,
    footerLocation: MENUS.FOOTER_LOCATION,
  };
};

SingleNews.query = gql`
  ${BlogInfoFragment}
  ${NavigationMenu.fragments.entry}
  query GetPost(
    $uri: ID!
    $headerLocation: MenuLocationEnum
    $footerLocation: MenuLocationEnum
  ) {
    post(id: $uri, idType: URI) {
